### PR TITLE
fix(core): name the module when metadata entries are invalid

### PR DESCRIPTION
**Pull request by thekhegay**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?
A custom provider whose `useClass`, `useFactory` or `useExisting` is `undefined` or `null`, or one with none of the `use*` properties, is silently skipped by `Module.addCustomProvider`, and bootstrap later fails with `TypeError: Cannot read properties of undefined (reading 'metatype')`. An `undefined` entry in `controllers` or `exports` fails with `reading 'name'` or `reading 'provide'`. None of these names the module or the token.

Issue Number: #17599 (#17600 made `null` behave like `undefined` but still falls through). Related: #16469, #12275

## What is the new behavior?
`addCustomProvider` throws an `InvalidProviderException` naming the module and the token, e.g. `The provider "X" in the AppModule "providers" array does not define a valid "useClass", "useValue", "useFactory" or "useExisting" property.`, with the likely causes. An `undefined` controller or export throws the same `CircularDependencyException` that `providers: [undefined]` already does. Valid metadata, including `useValue: undefined`, behaves as before.

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

## Other information
A request-scoped or transient `APP_GUARD` / `APP_INTERCEPTOR` with an undefined `useClass` or `useFactory` used to get through the scan and either boot without the enhancer or crash later with `reading 'isDependencyTreeStatic'`; it now fails at the scan with the message above.

### CODE REVIEW

File: packages/core/injector/module.ts [hunk 1]
```diff
@@ -318,6 +319,11 @@ export class Module {
       this.addCustomFactory(provider, collection, enhancerSubtype);
     } else if (this.isCustomUseExisting(provider)) {
       this.addCustomUseExisting(provider, collection, enhancerSubtype);
+    } else {
+      throw new InvalidProviderException(
+        (provider as ClassProvider).provide,
+        this.name,
+      );
```
Comment (micalevisk) [inline packages/core/injector/module.ts:326]: I ran an app-level probe against `master` and this branch: every provider shape that reaches this branch already crashed on `master` with `reading 'metatype'`, and `useValue: undefined` still boots. So no working app should notice this.

The one path where behaviour really changes is the one you called out in the description: `REQUEST`/`TRANSIENT`-scoped `APP_*` enhancers go through `addInjectable`, so on `master` `{ provide: APP_GUARD, scope: Scope.REQUEST, useClass: undefined }` boots silently when the app has no routes (and dies with `reading 'isDependencyTreeStatic'` as soon as it has one). Failing fast is the right call there.

Since that's the only path with a real behaviour change, please pin it with a spec through `addInjectable`. Something like this in the existing `describe('addInjectable')` block (I checked that it passes):

```ts
it('should throw when a custom injectable does not define a valid "use*" property', () => {
  expect(() =>
    moduleRef.addInjectable(
      { provide: 'test', useClass: undefined } as any,
      'guard',
    ),
  ).toThrow(InvalidProviderException);
});
```

File: packages/core/errors/messages.ts [hunk 2]
```diff
@@ -224,6 +224,16 @@ Potential causes:
...
 
+export const INVALID_PROVIDER_MESSAGE = (
+  token: InjectorDependency,
+  moduleName: string,
+) => `Nest cannot create the ${moduleName} instance.
+The provider ${getDependencyName(token, 'provider')} in the ${moduleName} "providers" array does not define a valid "useClass", "useValue", "useFactory" or "useExisting" property.
+
+Potential causes:
+- The value of "useClass", "useFactory" or "useExisting" is undefined at runtime, often because of a circular import between files. Check your import statements.
+- None of these properties is set, or the one that is set is null.`;
```
Comment (micalevisk) [inline packages/core/errors/messages.ts:235]: Wording and structure match `UNDEFINED_MODULE_MESSAGE` / `INVALID_MODULE_MESSAGE` nicely, and I checked it renders well for string, symbol and class tokens (`"X"`, `Symbol(SYM)`, `DenyGuard`).

Please add a `describe('INVALID_PROVIDER_MESSAGE')` to `packages/core/test/errors/test/messages.spec.ts`. The other builders in this file are asserted verbatim there (see `UNDEFINED_MODULE_EXCEPTION`), while the new spec in `module.spec.ts` only checks a substring with a string token — a symbol and a class token would be worth covering too.

FYI, non-blocking: for global enhancers the scanner rewrites the token before it gets here (`DependenciesScanner#insertProvider`), so the message reads `The provider "APP_GUARD (UUID: 97c1a45b18667b251c66f)" in the AppModule "providers" array…`. Still recognizable, so I wouldn't hold the PR for it.

File: packages/core/injector/container.ts [hunk 3]
```diff
@@ -299,6 +303,9 @@ export class NestContainer {
       throw new UnknownModuleException();
     }
     const moduleRef = this.modules.get(token)!;
+    if (isNil(controller)) {
+      throw new CircularDependencyException(moduleRef.metatype.name);
+    }
```
Comment (micalevisk) [inline packages/core/injector/container.ts:308]: Mirroring `addProvider` is the right call for this PR, and keeping the check after the `UnknownModuleException` one preserves the existing `addController(null!, 'TestModule')` spec 👍

Non-blocking, follow-up material: for `controllers` the `CircularDependencyException` text is a bit of a stretch. It tells the user to decorate "each side of a bidirectional relationship" with `forwardRef()`, but `forwardRef()` isn't resolved for `controllers` (nor `providers`) entries — the real cause is almost always a circular *file* import / barrel file or a typo, which is exactly what your new `INVALID_PROVIDER_MESSAGE` explains so well. It also doesn't say which array or index is the offender.

For the `imports` counterpart (#4586 / #4587) the reasoning was that circular deps always surface as `undefined`, never `null` (`null` = developer mistake), and the merged `UNDEFINED_MODULE_MESSAGE` lists both potential causes plus the index. `addProvider` (`!provider`) already conflates the two, so what you did is consistent with today's behaviour and I'm not asking to change it here — but the same treatment for `providers` / `controllers` / `exports` would be a welcome follow-up.

File: packages/core/test/injector/container.spec.ts [hunk 4]
```diff
@@ -33,6 +33,24 @@ describe('NestContainer', () => {
...
+  });
+
+  it('should "addExportedProviderOrModule" throw "CircularDependencyException" naming the module when exported item is nil', async () => {
+    const { moduleRef } = (await container.addModule(TestModule, []))!;
+    const addExport = () =>
+      container.addExportedProviderOrModule(undefined!, moduleRef.token);
+
+    expect(addExport).toThrow(CircularDependencyException);
+    expect(addExport).toThrow('inside TestModule');
+  });
```
Comment (micalevisk) [inline packages/core/test/injector/container.spec.ts]: The titles say *nil* but only `undefined` is exercised, while the guard is `isNil` (and on `master` `null` fails with a different `TypeError`: `Cannot read properties of null`). Please cover both — I checked that this passes:

```suggestion
  it.each([undefined, null])(
    'should "addController" throw "CircularDependencyException" naming the module when controller is %s',
    async nilValue => {
      const { moduleRef } = (await container.addModule(TestModule, []))!;
      const addController = () =>
        container.addController(nilValue!, moduleRef.token);

      expect(addController).toThrow(CircularDependencyException);
      expect(addController).toThrow('inside TestModule');
    },
  );

  it.each([undefined, null])(
    'should "addExportedProviderOrModule" throw "CircularDependencyException" naming the module when exported item is %s',
    async nilValue => {
      const { moduleRef } = (await container.addModule(TestModule, []))!;
      const addExport = () =>
        container.addExportedProviderOrModule(nilValue!, moduleRef.token);

      expect(addExport).toThrow(CircularDependencyException);
      expect(addExport).toThrow('inside TestModule');
    },
  );
```

**Review by micalevisk — CHANGES_REQUESTED** [review]
Thanks @thekhegay — nicely scoped, and the description is accurate down to the edge cases.

**What I checked**

- Ran an app-level probe (`Test.createTestingModule()` + an HTTP request) against `master` and this branch. Every shape that now throws `InvalidProviderException` already crashed on `master` with `reading 'metatype'`: nil `useClass` / `useFactory` / `useExisting`, only `provide`, string / symbol / class tokens, and static-scoped `APP_*`. `controllers: [undefined | null]`, `exports: [undefined]` and `exports: [forwardRef(() => undefined)]` went from raw `TypeError`s to a `CircularDependencyException` naming the module.
- `useValue: undefined` (#11129) still boots, and testing overrides are untouched since `Module#replace` goes through `mergeWith` rather than `addCustomProvider`.
- The only "boots today → throws now" case is the one you disclosed: a `REQUEST`/`TRANSIENT`-scoped `APP_*` enhancer with a nil `useClass`/`useFactory` **in an app with no routes**. As soon as there is a route it already crashed with `reading 'isDependencyTreeStatic'`. Failing fast beats silently dropping a global guard, so I agree with the "no breaking change" box.
- Unit specs and `oxlint` pass locally; the branch is up to date with `master`.

**Prior art** — this is consistent with how we've handled this family of errors before: #4587 (nil `imports`), #9506 (module name for `providers: [undefined]`), #17600 and #17612 were all merged as fixes. What got rejected in the past was *silently skipping* undefined providers (#9912), which is the opposite of what this PR does.

**Requested changes (details inline)** — the implementation itself looks good to me, I'd just like the new behaviour properly pinned by tests before this lands:

1. a spec through `addInjectable` — the one path where behaviour really changes
2. a verbatim `INVALID_PROVIDER_MESSAGE` case in `messages.spec.ts`
3. cover `null` in the two new container specs

Happy to approve once those three specs are in.

**Possible follow-up (not for this PR, non-blocking):** an `UNDEFINED_MODULE_MESSAGE`-style error (array name + index, both potential causes) for nil entries in `providers` / `controllers` / `exports`, instead of the `forwardRef()`-centric `CircularDependencyException` text.

**Comment by thekhegay** [comment]
Thanks for the thorough check @micalevisk

App-level probe is more than I did.

All three are in e3dd77e3:
- a spec through `addInjectable` in the existing `describe` block
- `describe('INVALID_PROVIDER_MESSAGE')` in `messages.spec.ts`, asserted verbatim for a string, a symbol and a class token
- your `it.each([undefined, null])` suggestion applied as is to both container specs

Full unit suite green (3039), oxlint and prettier clean. I also reverted three source files to master and re-ran: all 13 specs in those files fail without the change.

On the follow-up: I'd happy to do the `UNDEFINED_MODULE_MESSAGE`-style error (array name, index, both potential causes) for `providers` / `controllers` / `exports` in a separate PR, if that's wanted.
